config/settings.js

```diff
@@ -1,18 +1,70 @@
-module.exports = {
-    port: process.env.PORT || 8000,
-    env: process.env.ENV || 'production',
+const ALL = '*';
+const PROD = 'production';
+const DEV = 'development';
+const TEST = 'testing';
 
-    // Environment-dependent settings
-    development: {
-        db: {
+const env = process.env.ENV || PROD;
+
+const defaults = {};
+
+defaults[ALL] = {
+    port: 8000,
+    env: env,
+    db: {
+        options: {
             dialect: 'sqlite',
             storage: ':memory:'
         }
-    },
-    production: {
-        db: {
-            dialect: 'sqlite',
+    }
+};
+
+defaults[PROD] = {
+    db: {
+        options: {
             storage: 'db/database.sqlite'
         }
     }
 };
+
+defaults[DEV] = {
+};
+
+defaults[TEST] = {
+    db: {
+        options: {
+            logging: false
+        }
+    }
+};
+
+const fromEnvironment = {
+    port: process.env.PORT,
+    db: {
+        url: process.env.DATABASE_URL,
+        database: process.env.DATABASE_NAME,
+        username: process.env.DATABASE_USER,
+        password: process.env.DATABASE_PASSWORD
+    }
+}
+
+const deepAssign = (target, ... sources) => {
+    sources.forEach (source => {
+        Object.keys (source).forEach (key => {
+            if ( source[key] === undefined ) { return; }
+            if ( typeof target[key] === 'object' ) {
+                target[key] = deepAssign ({}, target[key], source[key]);
```

Comment:
I'm not sure if this is intentional. But if you have one source with a key assigned to a value (e.g. a string) it will overwrite your object of previous sources completely. I would raise an exception when the source and target types for `key` are not the same

`deepAssign({}, {'key': {'nested": 'true'}}, {'key': 'overwritten'}` would yield `{'key': 'overwritten}`

Reply:
I would also move this to a utility module. Instead of keeping it in the settings

Reply:
Blindly copying over when the types can not be merge (which in this case means it is not some sort of object) is intentional. That is what `Object.assign` would do in such a case, so i figured i'd mimic that. If you prefer i could change the behaviour. I don't know what you'd like more, staying close to `Object.assign` or throwing if there is a mismatch.

I'm a bit torn on putting the method in so utility module.

If you want a utility you can use something like https://github.com/KyleAMathews/deepmerge. I only made a simple version of this because I wanted to get rid of undefined entries. Maybe there is an implicit dependency on that project (or something similar) but  I did not want to introduce an explicit one just to merge these config objects.

I don't think this function is something that we want to reuse as it is right now. It does not handle arrays (here it doesn't need to, and will probably never need to). It probably does not handle other specific cases either.